src/feature/dirclient/dirclient.c

```diff
@@ -759,6 +760,10 @@ connection_dir_client_request_failed(dir_connection_t *conn)
     log_info(LD_DIR, "Giving up on downloading microdescriptors from "
              "directory server at '%s'; will retry", conn->base_.address);
     connection_dir_download_routerdesc_failed(conn);
+  } else if (TO_CONN(conn)->purpose == DIR_PURPOSE_UPLOAD_HSDESC &&
+             conn->hs_ident) {
+    log_info(LD_DIR, "Giving up on uploading HS descriptor. Retrying.");
+    hs_service_reupload_desc_to_dir(conn->hs_ident, conn->identity_digest);
```

Comment:
Hmm are we talking about infinite retries here? Could this be a problem?
Like, if the network goes down (e.g. mobile), will the service go mental?

Reply:
Yeah I thought of that as well and figured you would notice it and give me a good idea here lol...

Would be wise to avoid spamming to infinity the reupload to a single HSDir. We would need something like `last_hid_serv_request` used for clients. And avoid too many retries.

But there is probably an larger problem here which is that the service side does _not_ keep a state of its desc upload. So imagine all 6 upload failed and retried many times but then we stop, chances are that our network is down for instance. When it comes back up, we should notice that we never uploaded and do it again.

Else, the service thinks it made it happen and will wait between 60 and 120 min before retrying...

There is actually an argument now I think to extend this feature to correctly track our HSDir upload service side. What do you think?

Reply:
Hmm, you are right, I'm not sure what's the ideal behavior here. I'm generally afraid of this code path spinning in an andless loop when the network is down. Perhaps we can add a small delay when the network fails us before we retry. But not sure how hard that would be engineering wise :/